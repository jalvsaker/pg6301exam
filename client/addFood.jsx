import { useState } from "react";

export function AddFood({ reload }) {
  const [name, setName] = useState("");
  const [price, setPrice] = useState("");

  async function handleSubmit(e) {
    e.preventDefault();

    const body = { name, price };

    await fetch("/api/foods", {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(body),
    });

    reload();
  }

  return (
    <form onSubmit={handleSubmit}>
      <label>
        Name:
        <input
          type={"text"}
          value={name}
          onChange={(e) => setName(e.target.value)}
        />
      </label>
      <label>
        Price:
        <input
          type={"number"}
          value={price}
          onChange={(e) => setPrice(e.target.value)}
        />
      </label>
      <button>Submit</button>
    </form>
  );
}
