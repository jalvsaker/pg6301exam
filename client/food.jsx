export function Food({ food }) {
  return (
    <div>
      {food.name} - {food.price}
      <div>
        <input type="number" /> <button>Add to cart</button>
      </div>
      <hr />
    </div>
  );
}
