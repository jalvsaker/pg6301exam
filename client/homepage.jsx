export function Homepage() {
  return <h1>Hello world</h1>;
}
