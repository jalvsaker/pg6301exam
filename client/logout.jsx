import { useNavigate } from "react-router-dom";
import { useEffect } from "react";

export function Logout({ setUser }) {
  const navigate = useNavigate();

  useEffect(() => {
    (async () => {
      const res = await fetch("/api/login", { method: "delete" });
      if (res.ok) {
        setUser({});
        navigate("/");
      }
    })();
  }, []);

  return "Logging out...";
}