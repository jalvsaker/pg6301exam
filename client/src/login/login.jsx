import React from "react";
import { useState } from "react";
import { useNavigate } from "react-router-dom";

export function Login({ setUser }) {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");

  const navigate = useNavigate();

  async function handleSubmit(e) {
    e.preventDefault();

    const res = await fetch("/api/login", {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ username: username, password: password }),
    });

    if (res.ok) {
      setUser(await res.json());
      navigate("/");
    } else {
      alert(`${res.status} ${res.statusText}`);
    }
  }

  return (
    <form onSubmit={handleSubmit}>
      <h1>Log in:</h1>
      <div>
        <label>
          Username:{" "}
          <input
            type="text"
            value={username}
            onChange={(e) => setUsername(e.target.value)}
          />
        </label>
      </div>
      <div>
        <label>
          Password:{" "}
          <input
            type="password"
            value={password}
            onChange={(e) => setPassword(e.target.value)}
          />
        </label>
      </div>
      <button>Log in</button>
    </form>
  );
}
