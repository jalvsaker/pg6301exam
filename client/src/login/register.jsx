import React from "react";
import { useState } from "react";
import { useNavigate } from "react-router-dom";

export function Register({ setUser }) {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");

  const navigate = useNavigate();

  async function handleSubmit(e) {
    e.preventDefault();

    const res = await fetch("/api/login/new", {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ username: username, password: password }),
    });

    if (res.ok) {
      setUser(await res.json());
      navigate("/");
    } else {
      alert(`${res.status} ${res.statusText}`);
    }
  }

  return (
    <form onSubmit={handleSubmit}>
      <h1>Register new user:</h1>
      <div>
        <label>
          Username:{" "}
          <input
            type="text"
            value={username}
            onChange={(e) => setUsername(e.target.value)}
          />
        </label>
      </div>
      <div>
        <label>
          Password:{" "}
          <input
            type="password"
            value={password}
            onChange={(e) => setPassword(e.target.value)}
          />
        </label>
      </div>
      <button>Log in</button>
    </form>
  );
}
