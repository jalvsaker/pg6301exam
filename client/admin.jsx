import { useNavigate } from "react-router-dom";
import { useEffect, useState } from "react";
import { AddFood } from "./addFood";

export function Admin({ user }) {
  const navigate = useNavigate();

  if (!user.isAdmin) {
    navigate("/");
  }

  const [foods, setFoods] = useState([]);
  const [loading, setLoading] = useState(true);
  const [reloader, setReloader] = useState(false);

  useEffect(() => {
    (async () => {
      const res = await fetch("/api/foods");
      if (res.ok) {
        setFoods(await res.json());
        setLoading(false);
      }
    })();
  }, [reloader]);

  function reload() {
    setReloader((r) => !r);
  }

  if (loading) {
    return <div>Loading...</div>;
  }

  return (
    <>
      <h1>Admin panel</h1>

      <AddFood reload={reload} />

      {foods.map((food) => (
        <div key={food._id}>{food.name}</div>
      ))}
    </>
  );
}
