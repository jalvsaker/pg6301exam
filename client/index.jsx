import React from "react";
import { createRoot } from "react-dom/client";
import {
  BrowserRouter,
  Navigate,
  NavLink,
  Route,
  Routes,
} from "react-router-dom";
import { Homepage } from "./homepage";

function App() {
  return (
    <BrowserRouter>
      <nav>
        <NavLink to={"/"}>Home</NavLink>
        <NavLink to={"/test"}>Test</NavLink>
      </nav>
      <Routes>
        <Route path={"/"} element={<Homepage />} />
        <Route path={"/test"} element={<h2>test</h2>} />
        <Route path={"/*"} element={<Navigate to={"/"} />} />
      </Routes>
    </BrowserRouter>
  );
}

const root = createRoot(document.getElementById("root"));
root.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>
);
