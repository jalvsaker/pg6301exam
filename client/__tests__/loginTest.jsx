import { createRoot } from "react-dom/client";
import { act, Simulate } from "react-dom/test-utils";
import { MemoryRouter } from "react-router-dom";
import { Logout } from "../src/login/logout";
import React from "react";
import { Login } from "../src/login/login";
import { Register } from "../src/login/register";


describe("login test", function() {
  it("should show logout", async () => {
    global.fetch = jest.fn(() => {
      return {
        ok: false,
        json: () => {
          return {};
        },
      };
    });

    const element = document.createElement("div");
    const root = createRoot(element);

    const mock = jest.fn()

    await act(() => {
      root.render(
        <MemoryRouter>
          <Logout setUser={mock}/>
        </MemoryRouter>
      );
    });

    expect(element.innerHTML).toMatchSnapshot();
  });

  it("should submit login", async () => {
    global.fetch = jest.fn(() => {
      return {
        ok: true,
        json: () => {
          return {};
        },
      };
    });
    window.alert = jest.fn();

    const element = document.createElement("div");
    const root = createRoot(element);

    const mock = jest.fn();

    await act(() => {
      root.render(
        <MemoryRouter><Login setUser={mock}/></MemoryRouter>
      );
    });

    await act(() => {
      Simulate.change(element.querySelectorAll("input")[0], {
        target: { value: "username" },
      });
      Simulate.change(element.querySelectorAll("input")[1], {
        target: { value: "password" },
      });
    });

    await act(() => {
      Simulate.submit(element.querySelector("form"));
    });

    expect(element.innerHTML).toMatchSnapshot();
    expect(global.fetch).toBeCalled();
  });

  it("should fail to submit login", async () => {
    global.fetch = jest.fn(() => {
      return {
        ok: false,
        json: () => {
          return {};
        },
      };
    });
    window.alert = jest.fn();

    const element = document.createElement("div");
    const root = createRoot(element);

    const mock = jest.fn();

    await act(() => {
      root.render(
        <MemoryRouter><Login setUser={mock}/></MemoryRouter>
      );
    });

    await act(() => {
      Simulate.change(element.querySelectorAll("input")[0], {
        target: { value: "username" },
      });
      Simulate.change(element.querySelectorAll("input")[1], {
        target: { value: "password" },
      });
    });

    await act(() => {
      Simulate.submit(element.querySelector("form"));
    });

    expect(element.innerHTML).toMatchSnapshot();
    expect(global.fetch).toBeCalled();
    expect(window.alert).toBeCalled();
  });

  it("should submit register", async () => {
    global.fetch = jest.fn(() => {
      return {
        ok: true,
        json: () => {
          return {};
        },
      };
    });
    window.alert = jest.fn();

    const element = document.createElement("div");
    const root = createRoot(element);

    const mock = jest.fn();

    await act(() => {
      root.render(
        <MemoryRouter><Register setUser={mock}/></MemoryRouter>
      );
    });

    await act(() => {
      Simulate.change(element.querySelectorAll("input")[0], {
        target: { value: "username" },
      });
      Simulate.change(element.querySelectorAll("input")[1], {
        target: { value: "password" },
      });
    });

    await act(() => {
      Simulate.submit(element.querySelector("form"));
    });

    expect(element.innerHTML).toMatchSnapshot();
    expect(global.fetch).toBeCalled();
  });
  it("should fail to submit register", async () => {
    global.fetch = jest.fn(() => {
      return {
        ok: false,
        json: () => {
          return {};
        },
      };
    });
    window.alert = jest.fn();

    const element = document.createElement("div");
    const root = createRoot(element);

    const mock = jest.fn();

    await act(() => {
      root.render(
        <MemoryRouter><Register setUser={mock}/></MemoryRouter>
      );
    });

    await act(() => {
      Simulate.change(element.querySelectorAll("input")[0], {
        target: { value: "username" },
      });
      Simulate.change(element.querySelectorAll("input")[1], {
        target: { value: "password" },
      });
    });

    await act(() => {
      Simulate.submit(element.querySelector("form"));
    });

    expect(element.innerHTML).toMatchSnapshot();
    expect(global.fetch).toBeCalled();
    expect(window.alert).toBeCalled();
  });
});