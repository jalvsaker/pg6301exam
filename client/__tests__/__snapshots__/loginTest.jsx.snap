// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login test should show logout 1`] = `"Logging out..."`;

exports[`login test should submit login 1`] = `"<form><h1>Log in:</h1><div><label>Username: <input type="text" value="username"></label></div><div><label>Password: <input type="password" value="password"></label></div><button>Log in</button></form>"`;

exports[`login test should submit register 1`] = `"<form><h1>Register new user:</h1><div><label>Username: <input type="text" value="username"></label></div><div><label>Password: <input type="password" value="password"></label></div><button>Log in</button></form>"`;
